internal/querycoordv2/task/scheduler.go

```diff
@@ -0,0 +1,683 @@
+package task
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/querycoordv2/meta"
+	"github.com/milvus-io/milvus/internal/querycoordv2/session"
+	"github.com/milvus-io/milvus/internal/querycoordv2/utils"
+	. "github.com/milvus-io/milvus/internal/util/typeutil"
+	"go.uber.org/zap"
+)
+
+const (
+	taskPoolSize = 128
+
+	TaskTypeGrow Type = iota + 1
+	TaskTypeReduce
+	TaskTypeMove
+)
+
+var (
+	ErrConflictTaskExisted = errors.New("ConflictTaskExisted")
+
+	// The task is canceled or timeout
+	ErrTaskCanceled = errors.New("TaskCanceled")
+
+	// The target node is offline,
+	// or the target segment is not in TargetManager,
+	// or the target channel is not in TargetManager
+	ErrTaskStale = errors.New("TaskStale")
+
+	// No enough memory to load segment
+	ErrResourceNotEnough = errors.New("ResourceNotEnough")
+
+	ErrTaskQueueFull = errors.New("TaskQueueFull")
+)
+
+type Type int32
+
+type replicaSegmentIndex struct {
+	ReplicaID int64
+	SegmentID int64
+}
+
+type replicaChannelIndex struct {
+	ReplicaID int64
+	Channel   string
+}
+
+type taskQueue struct {
+	// TaskPriority -> Tasks
+	buckets [][]Task
+
+	cap int
+}
+
+func newTaskQueue(cap int) *taskQueue {
+	return &taskQueue{
+		buckets: make([][]Task, len(TaskPriorities)),
+
+		cap: cap,
+	}
+}
+
+func (queue *taskQueue) Len() int {
+	taskNum := 0
+	for _, tasks := range queue.buckets {
+		taskNum += len(tasks)
+	}
+
+	return taskNum
+}
+
+func (queue *taskQueue) Cap() int {
+	return queue.cap
+}
+
+func (queue *taskQueue) Add(task Task) bool {
+	if queue.Len() >= queue.Cap() {
+		return false
+	}
+
+	queue.buckets[task.Priority()] = append(queue.buckets[task.Priority()], task)
+	return true
+}
+
+func (queue *taskQueue) Remove(task Task) {
+	bucket := &queue.buckets[task.Priority()]
+
+	for i := range *bucket {
+		if (*bucket)[i].ID() == task.ID() {
+			*bucket = append((*bucket)[:i], (*bucket)[i+1:]...)
+			break
+		}
+	}
+}
+
+// Range iterates all tasks in the queue ordered by priority from high to low
+func (queue *taskQueue) Range(fn func(task Task) bool) {
+	for priority := len(queue.buckets) - 1; priority >= 0; priority-- {
+		for i := range queue.buckets[priority] {
+			if !fn(queue.buckets[priority][i]) {
+				return
+			}
+		}
+	}
+}
+
+type Scheduler interface {
+	Add(task Task) error
+	Dispatch(node int64)
+	RemoveByNode(node int64)
+	GetNodeSegmentDelta(nodeID int64) int
+	GetNodeChannelDelta(nodeID int64) int
+}
+
+type taskScheduler struct {
```

Comment:
it is not seperate by collection?

Reply:
This scheduler executes task within segment/channel level.